File: frontend/lib/api/progress.ts
import axios from 'axios';

const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api/v1';

const getAuthHeader = () => {
  const token = localStorage.getItem('token');
  return { Authorization: `Bearer ${token}` };
};

export const progressApi = {
  getCourseProgress: async (courseId: string) => {
    const response = await axios.get(`${API_URL}/progress/courses/${courseId}`, {
      headers: getAuthHeader()
    });
    return response.data;
  },

  updateCourseProgress: async (courseId: string, moduleId: string, data: any) => {
    const response = await axios.post(`${API_URL}/progress/courses/${courseId}/update`, {
      module_id: moduleId,
      ...data
    }, {
      headers: getAuthHeader()
    });
    return response.data;
  },

  getLearningPathProgress: async (pathId: string) => {
    const response = await axios.get(`${API_URL}/progress/learning-path/${pathId}`, {
      headers: getAuthHeader()
    });
    return response.data;
  },

  getOverallProgress: async () => {
    const response = await axios.get(`${API_URL}/progress/overall`, {
      headers: getAuthHeader()
    });
    return response.data;
  }
};

File: backend/app/api/v1/endpoints/progress.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from typing import List, Optional
from uuid import UUID
from datetime import datetime

from app.db.session import get_db
from app.core.deps import get_current_active_user
from app.db.models.user import User

router = APIRouter()


@router.get("/progress/courses/{course_id}")
def get_course_progress(
    course_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Get user's progress in a specific course."""
    return {
        "course_id": str(course_id),
        "user_id": str(current_user.id),
        "progress_percentage": 65,
        "completed_modules": 5,
        "total_modules": 8,
        "time_spent": 1200,  # minutes
        "last_accessed": datetime.utcnow().isoformat()
    }


@router.post("/progress/courses/{course_id}/update")
def update_course_progress(
    course_id: UUID,
    module_id: UUID,
    completed: bool,
    time_spent: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Update course progress."""
    return {
        "course_id": str(course_id),
        "module_id": str(module_id),
        "completed": completed,
        "time_spent": time_spent,
        "updated_at": datetime.utcnow().isoformat()
    }


@router.get("/progress/learning-path/{path_id}")
def get_learning_path_progress(
    path_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Get progress in a learning path."""
    return {
        "path_id": str(path_id),
        "user_id": str(current_user.id),
        "progress_percentage": 40,
        "completed_courses": 2,
        "total_courses": 5,
        "estimated_completion_date": (datetime.utcnow() + timedelta(days=30)).isoformat()
    }


@router.get("/progress/overall")
def get_overall_progress(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Get user's overall learning progress."""
    return {
        "user_id": str(current_user.id),
        "total_courses_enrolled": 12,
        "courses_completed": 8,
        "courses_in_progress": 4,
        "total_time_spent": 4800,  # minutes
        "average_completion_rate": 67,
        "certificates_earned": 8
    }
